mlir/lib/Quantum/Transforms/annotate_function.cpp

```diff
@@ -35,14 +37,34 @@ bool isAnnotated(FunctionOpInterface op, const char *attr)
 
 bool invalidGradientOperation(FunctionOpInterface op)
 {
-    auto res = op.walk([](Operation *o) {
-        if (dyn_cast<MeasureOp>(o) || dyn_cast<catalyst::CustomCallOp>(o) ||
-            dyn_cast<catalyst::CallbackCallOp>(o)) {
+    ModuleOp mod = op->getParentOfType<ModuleOp>();
+    auto res = op.walk([&](Operation *o) {
+        if (dyn_cast<MeasureOp>(o) || dyn_cast<catalyst::CustomCallOp>(o)) {
             return WalkResult::interrupt();
         }
-        else {
-            return WalkResult::advance();
+        else if (auto callbackCall = dyn_cast<catalyst::CallbackCallOp>(o)) {
+            bool isDifferentiated = false;
+            auto callee = callbackCall.getCalleeAttr();
+            auto callback =
+                SymbolTable::lookupNearestSymbolFrom<FunctionOpInterface>(callbackCall, callee);
+            bool inactive = callback.getResultTypes().empty();
+            if (inactive) {
+                return WalkResult::advance();
+            }
+            auto uses = *SymbolTable::getSymbolUses(callback, mod);
+            for (SymbolTable::SymbolUse use : uses) {
+                Operation *user = use.getUser();
+                if (auto customGrad = dyn_cast<catalyst::gradient::CustomGradOp>(user)) {
+                    isDifferentiated |= customGrad.getCalleeAttr() == callee;
```

Comment:
Why is the callee compared, don't we know that this CustomGradOp is already a user of the callback?

Reply:
`CustomGradOp` has three different symbols. The callee, the forward and reverse. Just because `CustomGradOp` uses the same symbol the callback uses, it doesn't mean that it is guaranteed to be in the callee.

Reply:
Makes sense, thanks!